### PR TITLE
mgn unlocking 

### PR DESCRIPTION
**Pull request by josojo**
Integrates the logic for MGN withdrawals and some spare tests

### CODE REVIEW

File: contracts/interfaces/IDutchExchange.sol [hunk 1]
```diff
@@ -1,7 +1,10 @@
 pragma solidity ^0.5.0;
 import "@gnosis.pm/mock-contract/contracts/MockContract.sol";
+import "@gnosis.pm/dx-contracts/contracts/TokenFRT.sol";
 
 contract IDutchExchange {
+    TokenFRT public frtToken;
```
Comment (fleupold) [inline contracts/interfaces/IDutchExchange.sol]: Why is this necessary?

File: test/dx_mgn_pool.js [hunk 2]
```diff
@@ -296,5 +311,43 @@ contract("DxMgnPool", (accounts) => {
 
       await truffleAssert.reverts(instance.withdrawMagnolia(), "MGN has not been unlocked, yet")
     })
+    it("withdrawals the right amount of MGN", async () => {
```
Comment (fleupold) [inline test/dx_mgn_pool.js]:
```suggestion
    it("withdraws the right amount of MGN", async () => {
```

File: test/dx_mgn_pool.js [hunk 3]
```diff
@@ -296,5 +311,43 @@ contract("DxMgnPool", (accounts) => {
...
+      await waitForNBlocks(101, accounts[0])
+
+      const balanceOf = mgnToken.contract.methods.balanceOf(accounts[0]).encodeABI()
+      await mgnTokenMock.givenMethodReturnUint(balanceOf, 100)      
+
+      await instance.withdrawMGN()
+      
+      assert.equal(await instance.totalMgn.call(), 100)
+
+      await depositTokenMock.givenAnyReturnBool(true)
```
Comment (fleupold) [inline test/dx_mgn_pool.js]: You already have `depositTokenMock.givenAnyReturnBool(true)` called above. No need to call it again.

File: test/dx_mgn_pool.js [hunk 4]
```diff
@@ -296,5 +311,43 @@ contract("DxMgnPool", (accounts) => {
...
+      await instance.withdrawDeposit()
+
+      await mgnTokenMock.givenAnyReturnBool(true)
+      await instance.withdrawMagnolia()
+
+      const mgnTransfer = mgnToken.contract.methods.transfer(accounts[0], 0).encodeABI()
```
Comment (fleupold) [inline test/dx_mgn_pool.js]: Shouldn't the expected amount here be 100 (the total amount of MGN)?

File: contracts/DxMgnPool.sol [hunk 5]
```diff
@@ -121,6 +122,18 @@ contract DxMgnPool {
         totalPoolSharesCummulative += totalPoolShares;
     }
 
+    function triggerMGNunlock() public{
+        require(currentState() == State.PoolingEnded, "Pooling period is not yet over.");
```
Comment (fleupold) [inline contracts/DxMgnPool.sol]: Shouldn't we also claim the remaining funds from the dx auctions here? Otherwise we have no logic to recover the funds from the last auction, before the pooling ends.

File: contracts/DxMgnPool.sol [hunk 6]
```diff
@@ -121,6 +122,18 @@ contract DxMgnPool {
...
+
+        mgnToken.unlockTokens();
+    }
+
+
+    function withdrawMGN() public {
+        mgnToken.withdrawUnlockedTokens();
```
Comment (fleupold) [inline contracts/DxMgnPool.sol]: For safety, should this also only be allowed when pooling ended?

File: contracts/DxMgnPool.sol [hunk 7]
```diff
@@ -146,6 +160,9 @@ contract DxMgnPool {
     
     function calculateClaimableMgn(Participation memory participation) private returns (uint) {
         uint duration = auctionCount - participation.startAuctionCount;
+        if (totalPoolSharesCummulative == 0) {
```
Comment (fleupold) [inline contracts/DxMgnPool.sol]: Is this necessary? If there was no totalPoolShareCummulative, this means there was no MGN minted, so it's fine to fail, isn't it?

Comment (josojo) [reply contracts/DxMgnPool.sol]: Yeah, it was just good for debugging. I hate reverts without error messages. /0 is one of them

File: test/dx_mgn_pool.js [hunk 8]
```diff
@@ -296,5 +311,41 @@ contract("DxMgnPool", (accounts) => {
...
+      const depositTokenMock = await MockContract.new()
+      const secondaryTokenMock = await MockContract.new()
+      const mgnTokenMock = await MockContract.new()
+      const mgnToken = await TokenFRT.deployed()
+      const dxMock = await MockContract.new()
+      const dx = await DutchExchange.new()
+      const poolingEndBlock = (await web3.eth.getBlockNumber()) + 100
+      const instance = await DxMgnPool.new(depositTokenMock.address, secondaryTokenMock.address, mgnTokenMock.address, dxMock.address, poolingEndBlock)
+      
+      await depositTokenMock.givenAnyReturnBool(true)
```
Comment (fleupold) [inline test/dx_mgn_pool.js]: I think a clean structure for a unit test is:
1) instantiation of component/dependent mocks
2) setting up the mocks with `given...` (grouped by mock)
3) interactions
4) assertions/expectation checking
